pkg/services/searchV2/manager.go

```diff
@@ -0,0 +1,299 @@
+				if err != nil {
+					m.logger.Error("Full re-indexing finished with error", "fullReIndexElapsed", time.Since(started), "error", err)
+				} else {
+					m.logger.Info("Full re-indexing finished", "fullReIndexElapsed", time.Since(started))
+				}
+				reIndexDoneCh <- lastIndexedEventID
+			}()
+		case lastIndexedEventID := <-reIndexDoneCh:
+			// Asynchronous re-indexing is finished. Set lastEventID to the value which
+			// was actual at the re-indexing start – so that we could re-apply all the
+			// events happened during async index build process and make sure it's consistent.
+			if lastEventID != lastIndexedEventID {
+				m.logger.Info("Re-apply event ID to last indexed", "currentEventID", lastEventID, "lastIndexedEventID", lastIndexedEventID)
+				lastEventID = lastIndexedEventID
+				// Apply events immediately.
+				partialUpdateTimer.Reset(0)
+			}
+			fullReIndexTimer.Reset(reIndexInterval)
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+	}
+}
+
+func (m *orgIndexManager) buildInitialIndexes(ctx context.Context, orgIDs []int64) error {
+	started := time.Now()
+	m.logger.Info("Start building in-memory indexes")
+	for _, orgID := range orgIDs {
+		err := m.buildInitialIndex(ctx, orgID)
+		if err != nil {
+			return fmt.Errorf("can't build initial dashboard search index for org %d: %w", orgID, err)
+		}
+	}
+	m.logger.Info("Finish building in-memory indexes", "elapsed", time.Since(started))
+	return nil
+}
+
+func (m *orgIndexManager) buildInitialIndex(ctx context.Context, orgID int64) error {
+	started := time.Now()
+	ctx, cancel := context.WithTimeout(ctx, time.Minute)
+	defer cancel()
+
+	m.logger.Info("Start building org index", "orgId", orgID)
+	index, err := m.indexFactory(ctx, orgID, nil)
+	if err != nil {
+		return fmt.Errorf("error building index: %w", err)
+	}
+	m.logger.Info("Finish building org index", "elapsed", time.Since(started), "name", m.name, "orgId", orgID)
+
+	m.mu.Lock()
+	m.perOrgIndex[orgID] = index
+	m.mu.Unlock()
+	return nil
+}
+
+func (m *orgIndexManager) reIndexExisting(ctx context.Context) error {
+	m.mu.RLock()
+	orgIDs := make([]int64, 0, len(m.perOrgIndex))
+	for orgID := range m.perOrgIndex {
+		orgIDs = append(orgIDs, orgID)
+	}
+	m.mu.RUnlock()
+
+	for _, orgID := range orgIDs {
+		m.mu.Lock()
+		i, ok := m.perOrgIndex[orgID]
+		if !ok {
+			// Skip event for org not yet indexed.
+			m.mu.Unlock()
+			continue
+		}
+		m.mu.Unlock()
+		err := i.ReIndex(ctx)
+		if err != nil {
+			return fmt.Errorf("error re-indexing for organization %d: %w", orgID, err)
+		}
+	}
+	return nil
+}
+
+func (m *orgIndexManager) applyIndexUpdates(ctx context.Context, lastEventID int64) int64 {
+	events, err := m.eventStore.GetAllEventsAfter(context.Background(), lastEventID)
+	if err != nil {
+		m.logger.Error("can't load events", "error", err)
+		return lastEventID
+	}
+	if len(events) == 0 {
+		return lastEventID
+	}
+	started := time.Now()
+
+	resourceEvents, err := store.GetResourceEvents(events)
```

Comment:
what is the conceptual difference between a `ResourceEvent` and an `EntityEvent`? Should we instead add a function like 
```
func parseEntityId(entityId string) (struct{
   storage: string, 
   orgId: int64, 
   uid: string, 
   kind: EntityType
})
```
?

Reply:
I refactored things a bit, the idea was to avoid parsing entity id in searchV2 code. And avoid having string entity ID in search v2 code. But I think we still need to have some new struct to wrap both ID, EntityEventType and ParsedEntityID. Since it differs from struct used for a DB model.